### PR TITLE
Refactor python test framework

### PR DESCRIPTION
**Pull request by patricklodder**
Pulled in from bitcoin because it's a dependency for #976

### CODE REVIEW

File: qa/rpc-tests/util.py [hunk 1]
```diff
@@ -88,8 +88,8 @@ def initialize_chain(test_dir):
             args = [ "dogecoind", "-keypool=1", "-datadir="+datadir, "-discover=0" ]
             if i > 0:
                 args.append("-connect=127.0.0.1:"+str(p2p_port(0)))
-            dogecoind_processes.append(subprocess.Popen(args))
-            subprocess.check_call([ "dogecoin-cli", "-datadir="+datadir,
+            dogecoind_processes[i] = subprocess.Popen(args)
+            subprocess.check_call([ "bitcoin-cli", "-datadir="+datadir,
```
Comment (rnicoll) [inline qa/rpc-tests/util.py]: Needs to be dogecoin-cli

**Comment by patricklodder** [comment]
fixed the issue re: missed dogecoin-cli line

**Comment by rnicoll** [comment]
Not sure it worked before, but definitely doesn't seem happy now:

~/dogecoin$ python qa/rpc-tests/listtransactions.py
Initializing test directory /tmp/testCITuDx
Unexpected exception caught during testing: [Errno 2] No such file or directory: 'cache/node1'
  File "/home/jrn/dogecoin/qa/rpc-tests/test_framework.py", line 63, in main
    self.setup_chain(self.options.tmpdir)
  File "/home/jrn/dogecoin/qa/rpc-tests/test_framework.py", line 33, in setup_chain
    initialize_chain(tmp_directory)
  File "/home/jrn/dogecoin/qa/rpc-tests/util.py", line 122, in initialize_chain
    shutil.copytree(from_dir, to_dir)
  File "/usr/lib/python2.7/shutil.py", line 171, in copytree
    names = os.listdir(src)
Cleaning up
Failed

**Comment by patricklodder** [comment]
It only works when I:
```
cd qa/rpc-tests
python listtransactions.py
cd ../..
```

If this is not fixed by the end of our sync list, I'll fix that myself.

**Pull request merged**